core/embed/trezorhal/stm32f4/systask.c

```diff
@@ -0,0 +1,601 @@
+    pminfo->privileged = privileged;
+    pminfo->fault.sp = sp;
+#if !(defined(__ARM_ARCH_8M_MAIN__) || defined(__ARM_ARCH_8M_BASE__))
+    pminfo->fault.sp_lim = task->sp_lim;
+#endif
+    pminfo->fault.irqn = (__get_IPSR() & IPSR_ISR_Msk) - 16;
+    pminfo->fault.cfsr = SCB->CFSR;
+    pminfo->fault.mmfar = SCB->MMFAR;
+    pminfo->fault.bfar = SCB->BFAR;
+    pminfo->fault.hfsr = SCB->HFSR;
+  }
+
+  systask_kill(task);
+
+  mpu_restore(mpu_mode);
+}
+
+// C part of PendSV handler that switches tasks
+//
+// `sp` is the stack pointer of the current task
+// `sp_lim` is the stack pointer limit of the current task
+// `exc_return` is the execution state of the current task
+//
+// Returns the context struct of the next task
+__attribute((no_stack_protector, used)) static uint32_t scheduler_pendsv(
+    uint32_t sp, uint32_t sp_lim, uint32_t exc_return) {
+  systask_scheduler_t* scheduler = &g_systask_scheduler;
+
+  // Save the current task context
+  systask_t* prev_task = scheduler->active_task;
+  prev_task->sp = sp;
+#if defined(__ARM_ARCH_8M_MAIN__) || defined(__ARM_ARCH_8M_BASE__)
+  // sp_lim is not valid on ARMv7-M
+  prev_task->sp_lim = sp_lim;
+#endif
+  prev_task->exc_return = exc_return;
+  prev_task->mpu_mode = mpu_get_mode();
+
+  // Switch to the next task
+  scheduler->active_task = scheduler->waiting_task;
+
+  // Load the scheduled task context
+  systask_t* next_task = scheduler->active_task;
+
+  // Set task privilege level
+  uint32_t control = __get_CONTROL();
+  if (next_task == &scheduler->kernel_task) {
+    control &= ~CONTROL_nPRIV_Msk;
+  } else {
+    control |= CONTROL_nPRIV_Msk;
+  }
+  __set_CONTROL(control);
+
+  // Setup the MPU for the new task
+  mpu_reconfig(next_task->mpu_mode);
+
+  return (uint32_t)next_task;
+}
+
+__attribute__((naked, no_stack_protector)) void PendSV_Handler(void) {
+  __asm__ volatile(
+      "LDR     R0, =%[active_task] \n"
+      "LDR     R1, =%[waiting_task]\n"
+      "CMP     R0, R1              \n"
+      "BEQ     3f                  \n"  // No task switch needed
+
+      "LDR     R0, [R0]            \n"  // R0 =  active_task
+      "LDR     R0, [R0, #12]       \n"  // R0 =  active_task->killed
+      "CMP     R0, #0              \n"
+      "BEQ     1f                  \n"  // =0 => normal processing
+
+      // We are switching from a killed task to the kernel task.
+      // Since the reason might be a stack overflow, we must not
+      // attempt to save the task context.
+
+      "LDR     R1, = 0xE000EF34    \n"  // FPU->FPCCR
+      "LDR     R0, [R1]            \n"
+      "BIC     R0, R0, #1          \n"  // Clear LSPACT to suppress later lazy
+      "STR     R0, [R1]            \n"  // stacking to the killed task stack
+
+      "MOV     R0, #0              \n"  // Skip context save
+      "MOV     R1, R0              \n"  //
+      "MOV     R2, R0              \n"  //
+      "B       2f                  \n"  //
+
+      "1:                          \n"
```

Comment:
could we use better labels then 1,2,3 in here?

Reply:
Yes, I would like to as well. Unfortunately, these numbered symbols are the only way to keep them local to the assembly block. We can use normal labels, but these symbols will be visible outside the scope of this block. I’m not sure which is worse.